drivers/can/can_rcar.c

```diff
@@ -0,0 +1,1044 @@
+				   const struct can_timing *timing)
+{
+	uint32_t bcr;
+
+	bcr = RCAR_CAN_BCR_TSEG1(timing->phase_seg1 + timing->prop_seg - 1) |
+	      RCAR_CAN_BCR_BPR(timing->prescaler - 1) |
+	      RCAR_CAN_BCR_SJW(timing->sjw - 1) |
+	      RCAR_CAN_BCR_TSEG2(timing->phase_seg2 - 1);
+
+	/* Don't overwrite CLKR with 32-bit BCR access; CLKR has 8-bit access.
+	 * All the registers are big-endian but they get byte-swapped on 32-bit
+	 * read/write (but not on 8-bit, contrary to the manuals)...
+	 */
+	sys_write32((bcr << 8) | RCAR_CAN_CLKR_CLKP2,
+		    config->reg_addr + RCAR_CAN_BCR);
+}
+
+int can_rcar_set_timing(const struct device *dev,
+			const struct can_timing *timing,
+			const struct can_timing *timing_data)
+{
+	const struct can_rcar_cfg *config = DEV_CAN_CFG(dev);
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+	int ret = 0;
+
+	ARG_UNUSED(timing_data);
+
+	k_mutex_lock(&data->inst_mutex, K_FOREVER);
+
+	/* Changing bittiming should be done in reset mode */
+	ret = can_rcar_enter_reset_mode(config, true);
+	if (ret != 0) {
+		goto unlock;
+	}
+
+	can_rcar_set_bittiming(config, timing);
+
+	/* Go back to operation mode */
+	ret = can_rcar_enter_operation_mode(config);
+
+unlock:
+	k_mutex_unlock(&data->inst_mutex);
+	return ret;
+}
+
+static void can_rcar_register_state_change_isr(const struct device *dev,
+					       can_state_change_isr_t isr)
+{
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+
+	data->state_change_isr = isr;
+}
+
+static enum can_state can_rcar_get_state(const struct device *dev,
+					 struct can_bus_err_cnt *err_cnt)
+{
+	const struct can_rcar_cfg *config = DEV_CAN_CFG(dev);
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+
+	if (err_cnt != NULL) {
+		can_rcar_get_error_count(config, err_cnt);
+	}
+	return data->state;
+}
+
+#ifndef CONFIG_CAN_AUTO_BUS_OFF_RECOVERY
+int can_rcar_recover(const struct device *dev, k_timeout_t timeout)
+{
+	return 0;
+}
+#endif /* CONFIG_CAN_AUTO_BUS_OFF_RECOVERY */
+
+int can_rcar_send(const struct device *dev, const struct zcan_frame *msg,
+		  k_timeout_t timeout, can_tx_callback_t callback,
+		  void *callback_arg)
+{
+	const struct can_rcar_cfg *config = DEV_CAN_CFG(dev);
+	struct can_rcar_data *data = DEV_CAN_DATA(dev);
+	struct can_rcar_tx_cb *tx_cb;
+	uint32_t identifier;
+	int i;
+
+	LOG_DBG("Sending %d bytes on %s. "
+		"Id: 0x%x, "
+		"ID type: %s, "
+		"Remote Frame: %s"
+		, msg->dlc, dev->name
+		, msg->id
+		, msg->id_type == CAN_STANDARD_IDENTIFIER ?
+		"standard" : "extended"
+		, msg->rtr == CAN_DATAFRAME ? "no" : "yes");
+
+	__ASSERT(msg->dlc == 0U || msg->data != NULL, "Dataptr is null");
+
+	if (msg->dlc > CAN_MAX_DLC) {
+		LOG_ERR("DLC of %d exceeds maximum (%d)",
+			msg->dlc, CAN_MAX_DLC);
+		return CAN_TX_EINVAL;
+	}
+
+	/* Wait for a slot into the tx FIFO */
+	if (k_sem_take(&data->tx_sem, timeout) != 0) {
+		return CAN_TIMEOUT;
+	}
+
+	k_mutex_lock(&data->inst_mutex, K_FOREVER);
+	tx_cb = &data->tx_cb[data->tx_head % RCAR_CAN_FIFO_DEPTH];
```

Comment:
This only works if RCAR_CAN_FIFO_DEPTH is a power of two. Otherwise you wil have a leap when the tx_head overflows

Reply:
that's true, this implementation looks a bit fragile, will try to find another trick.

Reply:
Ok did a second implementation when sending a can frame tx_head is still incremented but set to 0 in case it reaches RCAR_CAN_FIFO_DEPTH, the can_rcar_send function also increments data->tx_unsent, which is used by the tx isr handler, to know the fifo depth. That way it's easier to read and robust to any RCAR_CAN_FIFO_DEPTH value.